keymanapp/keyman
# fix(android/engine): Display longpress keys during a Move gesture

Addresses #5301 

~This PR is initially in Draft for @MakaraSok to characterize the behavior of 14.0 compared to this one.~
Currently, if a user starts a longpress gesture but moves off the initial base key, the longpress (subkeys) never show.
This is because the Keyman Engine for Android's use of GestureDetector isn't handling ACTION_MOVE when it interrupts ACTION_LONGPRESS.

This PR handles ACTION_MOVE during a longpress gesture to display subkeys (if available).

In talking with @mcdurdin , he said once longpress keys are displayed, it's ok to keep displaying the initial keys while the longpress is held. (We currently don't need to refresh to other longpress keys as the user drags to other keys)

KeymanWeb is also changed so the base key isn't updated if there's subkeys are being displayed. 
Previously, moving during the longpress would cause to longpress keys to update for the the final base key.
With this change, the longpress keys are locked into the original base key.

## User Testing
This needs to be tested on more than one device type

### Device Types
* **GROUP_WEB:** - Tests KeymanWeb desktop in mobile view
1. Load the PR build of KeymanWeb Test Home --> Manual Test pages --> Unminified KeymanWeb
2. Set the browser page to mobile viewport and hard refresh the page
3. Click the globe button on the keyboard and switch to Swedish (uses sil_euro_latin)
* **GROUP_ANDROID:** - Tests Android app
* **GROUP_IOS:** - Tests iOS app

### Tests
* **TEST_TOUCH_KEYS** - Tests the behavior of popup and longpress keys
1. Load the corresponding Keyman PR build of the group being tested
4. Enable Keyman as a system keyboard
5. Launch a separate app (e.g. Chrome or Safari) and select a field to start typing
6. With the default sil_euro_latin keyboard, select the symbolic layer
7. Short press a normal key <kbd>(</kbd>and release - verify `(` is output
8. Longpress a normal key <kbd>(</kbd>release - verify `(` is output
9. Longpress a key that contains longpress keys, and when the longpress keys display, move to an available longpress key and release - verify the selected longpress key is output
10. Longpress and move - Case 1:
    a. Longpress a key, but immediately move off the base key before the longpress registers. 
    b. While holding the longpress, verify longpress keys for the original base key are displayed
    c. Move to an available longpress key and release - verify the selected longpress key is output
11. Longpress and move - Case 2:
    a. Longpress a key, but immediately move off the base key before the longpress registers. 
    b. While holding the longpress, hold on a separate base key for a few seconds - verify longpress keys for the original base key are displayed
    c. Move to an available longpress key and release - verify the selected longpress key is output
12. Longpress and move - Case 3:
    a. Longpress a key, but immediately move off the base key before the longpress registers. 
    b. While holding the longpress, hold on a separate base key for a few seconds - verify longpress keys for the original base key are displayed
    c. Hold on the separate base key and release - verify nothing is output (only the longpress keys are eligible selections)
